pkg/tun/tcp/packet.go

```diff
@@ -0,0 +1,105 @@
+package tcp
+
+import (
+	"bytes"
+	"fmt"
+	"net"
+
+	"golang.org/x/sys/unix"
+
+	"github.com/telepresenceio/telepresence/v2/pkg/tun/buffer"
+	"github.com/telepresenceio/telepresence/v2/pkg/tun/ip"
+)
+
+type Packet interface {
+	ip.Packet
+	Header() Header
+	PayloadLen() int
+	Reset() Packet
+}
+
+type packet struct {
+	ipHdr   ip.Header
+	data    *buffer.Data
+	withAck bool
+}
+
+func PacketFromData(ipHdr ip.Header, data *buffer.Data) Packet {
+	return &packet{ipHdr: ipHdr, data: data}
+}
+
+func NewPacket(ipPayloadLen int, src, dst net.IP, withAck bool) Packet {
+	pkt := &packet{withAck: withAck}
+	ip.InitPacket(pkt, ipPayloadLen, src, dst)
+	return pkt
+}
+
+func (p *packet) IPHeader() ip.Header {
+	return p.ipHdr
+}
+
+func (p *packet) Data() *buffer.Data {
+	return p.data
+}
+
+func (p *packet) SetDataAndIPHeader(data *buffer.Data, ipHdr ip.Header) {
+	p.ipHdr = ipHdr
+	p.data = data
+}
+
+func (p *packet) SoftRelease() {
+	if !p.withAck {
+		p.Release()
+	}
+}
+
+func (p *packet) Release() {
+	buffer.DataPool.Put(p.data)
+}
+
+func (p *packet) Header() Header {
+	return p.IPHeader().Payload()
+}
+
+func (p *packet) PayloadLen() int {
+	return p.IPHeader().PayloadLen() - p.Header().DataOffset()*4
+}
+
+func (p *packet) String() string {
+	b := bytes.Buffer{}
```

Comment:
`strings.Builder` would be better for this case, since it would avoid an alloc+copy when you call `.String()`.

Reply:
The problem with `strings.Builder` is that it doesn't allow `Truncate()` which complicates the `AppendFlags()` function. Also, performance it not an issue for this function as it is only used for debugging.